programs/bpf_loader/src/syscalls.rs

```diff
@@ -934,901 +1150,1892 @@ impl<'a> SyscallObject<BpfError> for SyscallSha256<'a> {
-    })?;
-
-    Ok(SUCCESS)
+/// BIGNUM data cost
+const BIGNUM_WORDCOST: f64 = 6.0;
+/// BIGNUM cost divisor
+const BIGNUM_WORDCOST_DIVISOR: f64 = 32.0;
+/// Bignum data cost calculator
+macro_rules! calc_bignum_cost {
+    ($input:expr) => {
+        (BIGNUM_WORDCOST * ($input / BIGNUM_WORDCOST_DIVISOR).ceil()) as u64
+    };
 }
 
-/// Get a Clock sysvar
-struct SyscallGetClockSysvar<'a> {
-    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+/// BIGNUM sol_bignum_new
+pub struct SyscallBigNumNew<'a> {
+    cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
     loader_id: &'a Pubkey,
 }
-impl<'a> SyscallObject<BpfError> for SyscallGetClockSysvar<'a> {
+impl<'a> SyscallObject<BpfError> for SyscallBigNumNew<'a> {
     fn call(
         &mut self,
-        var_addr: u64,
+        bn_addr: u64,
         _arg2: u64,
         _arg3: u64,
         _arg4: u64,
         _arg5: u64,
         memory_mapping: &MemoryMapping,
         result: &mut Result<u64, EbpfError<BpfError>>,
     ) {
-        *result = get_sysvar::<Clock>(
-            &sysvar::clock::id(),
-            var_addr,
-            self.loader_id,
-            memory_mapping,
-            self.invoke_context.clone(),
+        let big_number = question_mark!(
+            translate_type_mut::<u64>(memory_mapping, bn_addr, self.loader_id, true),
+            result
         );
+        let bbox = Box::new(BigNum::new().unwrap());
+        let rwptr = Box::into_raw(bbox);
+        let bignum_ptr = rwptr as u64;
+        *big_number = bignum_ptr;
+        question_mark!(self.compute_meter.consume(self.cost), result);
+        *result = Ok(0)
     }
 }
-/// Get a EpochSchedule sysvar
-struct SyscallGetEpochScheduleSysvar<'a> {
-    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+
+pub struct SyscallBigNumFromU32<'a> {
+    cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
     loader_id: &'a Pubkey,
 }
-impl<'a> SyscallObject<BpfError> for SyscallGetEpochScheduleSysvar<'a> {
+impl<'a> SyscallObject<BpfError> for SyscallBigNumFromU32<'a> {
     fn call(
         &mut self,
-        var_addr: u64,
-        _arg2: u64,
+        bn_addr: u64,
+        u32_val: u64,
         _arg3: u64,
         _arg4: u64,
         _arg5: u64,
         memory_mapping: &MemoryMapping,
         result: &mut Result<u64, EbpfError<BpfError>>,
     ) {
-        *result = get_sysvar::<EpochSchedule>(
-            &sysvar::epoch_schedule::id(),
-            var_addr,
-            self.loader_id,
-            memory_mapping,
-            self.invoke_context.clone(),
+        let big_number = question_mark!(
+            translate_type_mut::<u64>(memory_mapping, bn_addr, self.loader_id, true),
+            result
         );
+        let bbox = Box::new(BigNum::from_u32(u32_val as u32).unwrap());
+        let bytes = bbox.num_bytes() as f64;
+        let rwptr = Box::into_raw(bbox);
+        let bignum_ptr = rwptr as u64;
+        *big_number = bignum_ptr;
+        question_mark!(
```

Comment:
Should consume first so that the call fails before doing any work if the caller doesn't have enough units left

Reply:
Moved consumes to early as possible depending on when size is available to calculate byte costs

Reply:
Fixed in 03af7519e1ca7a29ed2ee25c2cd6b31e0d4e99fd